src/librbd/ImageWatcher.cc

```diff
@@ -575,6 +632,98 @@ int ImageWatcher<I>::prepare_async_request(const AsyncRequestId& async_request_i
   return 0;
 }
 
+template <typename I>
+Context *ImageWatcher<I>::prepare_quiesce_request(
+    const AsyncRequestId &request, C_NotifyAck *ack_ctx) {
+  std::unique_lock l{m_async_request_lock};
+
+  if (m_async_pending.count(request) != 0) {
+    auto it = m_async_requests.find(request);
+    if (it != m_async_requests.end()) {
+      delete it->second.first;
+      it->second.first = ack_ctx;
+    } else {
+      m_task_finisher->queue(new C_ResponseMessage(ack_ctx), -ESTALE);
+    }
+    return nullptr;
+  }
+
+  m_async_pending.insert(request);
+  m_async_requests[request] = AsyncRequest(ack_ctx, nullptr);
+  m_async_op_tracker.start_op();
+
+  auto unquiesce_ctx = new LambdaContext(
+    [this, request](int r) {
+      if (r == 0) {
+        ldout(m_image_ctx.cct, 10) << this << " quiesce request " << request
+                                   << " timed out" << dendl;
+      }
+
+      auto on_finish = new LambdaContext(
+        [this, request](int r) {
+          std::unique_lock async_request_locker{m_async_request_lock};
+          m_async_pending.erase(request);
+        });
+
+      m_image_ctx.state->notify_unquiesce(on_finish);
+    });
+
+
+  return new LambdaContext(
+    [this, request, unquiesce_ctx](int r) {
+      ceph_assert(r == 0);
+      {
+        std::unique_lock locker{m_async_request_lock};
+        if (m_async_pending.count(request) == 0) {
+          delete unquiesce_ctx;
+          r = -ESTALE;
+        }
+      }
+
+      if (r == 0) {
+        auto attempts = m_image_ctx.config.template get_val<uint64_t>(
+          "rbd_quiesce_notification_attempts");
+        auto delay = 2 * attempts * watcher::Notifier::NOTIFY_TIMEOUT / 1000;
+
+        m_task_finisher->add_event_after(Task(TASK_CODE_QUIESCE, request),
```

Comment:
Nit: we will want to reset the timeout if we are still receiving quiesce pings (i.e. the timeout should be 2x the notify resend rate).